### DISCUSSION TITLE
FragAttacks

### DISCUSSION THREAD

**Discussion by nagyrobi** [HASwitchPlate/openHASP#182] [General]
@fvanroie please have a look at the latest Wifi vulnerabilities discussion on Tasmota:
https://github.com/arendst/Tasmota/discussions/12104
How deep should we be concerned?

Eg. I'm thinking of [this](https://www.fragattacks.com/):
_How can the adversary construct unencrypted Wi-Fi frames so they are accepted by a vulnerable device? First, certain Wi-Fi devices accept any unencrypted frame even when connected to a protected Wi-Fi network. This means the attacker doesn't have to do anything special! Two of out of four tested home routers were affected by this vulnerability, several internet-of-things devices were affected, and some smartphones were affected._

**Comment by fvanroie** [comment]
Espressif released an update to ESP-IDF which includes patches for the FragAttacks CVEs. See their [Security Advisory](https://www.espressif.com/sites/default/files/advisory_downloads/AR2021-003%20Security%20Advisory%20for%20WFA%20vulnerability.pdf).
The fixes have been added to ESP-IDF versions:
- Master (ef127ab9 )
- Release v4.3.1 (46144f70 )
- Release v4.2.2 (60ccb3fe )
- Release v4.1.2 (97c8be71 )
- Release v4.0.4 (7504329e )
- Release v3.3.6 (b403b0db )

At the moment, Arduino core 1.0.6 still uses ESP-IDF 3.3.5. We will update it as soon as a new core become available.

**Comment by nagyrobi** [comment]
Additionally we should consider configuration options to disable telnet/http (only MQTT-operation).

Also implement some extra security like encryption for MQTT client like SSL/TLS as HA [already supports](https://www.home-assistant.io/docs/mqtt/certificate/) it.

**Reply by nagyrobi** [reply]
Some [tips](https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/) from Tasmota.

**Reply by fvanroie** [reply]
To remove the Telnet and Http servers use in the build flags:
```
    -D HASP_USE_TELNET=0
    -D HASP_USE_HTTP=0
```
or specify an http username and password to block access.

**Reply by nagyrobi** [reply]
I thought of a configuration at runtime. 

One could still need the http to update/backup jsonl, upload files etc. but could disable it after that.

Eg: An mqtt command which could simply put a flag in the config not to start http or telnet at boot. At next boot, if the corresponding flag value is true, skip starting the http or telnet service. The flag could be reset though mqtt to default false, after a new reboot the required services would start.

`hasp/plate/command/disablehttp 1` etc

**Reply by fvanroie** [reply]
How about the username and password...

**Reply by nagyrobi** [reply]
It's plaintext...

**Reply by fvanroie** [reply]
gotcha

**Reply by fvanroie** [reply]
How about:

```python
service start telnet
TELN: Telnet console started
service stop telnet
TELN: Closing session from 192.168.0.123
service start http
HTTP: Started @ http://192.168.0.5
service stop http
HTTP: Stopped
```

**Reply by nagyrobi** [reply]
1. How do you run these after you stopped telnet? (no serial access in the wall)
2. The goal would be to avoid starting at boot (think of power outage). Ubuntu-like equivalent for the above `service disable telnet`

**Reply by fvanroie** [reply]
Just send it as a service command with payload `start/stop http/telnet`

**Reply by nagyrobi** [reply]
can this be added to the [docs](https://haswitchplate.github.io/openHASP-docs/0.6/commands/)?

**Reply by nagyrobi** [reply]
[19:28:06.212][92188/112580 16][22640/24860  9] MSGR: service=stop http
[19:28:06.230][92188/112652 18][22640/24860  9] MSGR: Command 'service' not found => stop http

was this just an idea?

**Reply by fvanroie** [reply]
It was a quick test to see if/how it could work. I'm OK with implementing it, but not sure about the command structure yet:

`service start|stop <name>`
`<name> start|stop`
`start|stop <name>`
`...`

**Reply by nagyrobi** [reply]
A kind of autoexec.bat implementation would be needed to run the commands at boot.

**Reply by fvanroie** [reply]
> How about:
> 
> ```python
> service start telnet
> TELN: Telnet console started
> service stop telnet
> TELN: Closing session from 192.168.0.123
> service start http
> HTTP: Started @ http://192.168.0.5
> service stop http
> HTTP: Stopped
> ```

Can be tested in master

**Reply by nagyrobi** [reply]
Works:
publish to `hasp/<your plate>/command/service` the string `stop telnet` etc

**Comment by nagyrobi** [comment]
Not FragAttacks, but still interesting read:
https://github.com/arendst/Tasmota/issues/6767

**Comment by fvanroie** [comment]
FragAttack has been fixed in [arduino-esp32 v2.0.0](https://github.com/espressif/arduino-esp32/issues/5351#event-5343676499)

Now the [platformio framework](https://github.com/platformio/platform-espressif32/issues/619) needs to update too, so we can use it.

**Reply by fvanroie** [reply]
Upvote the PIO link because we will be needing arduino-esp32 v2.0.0 for ESP32-S2 support too...

**Reply by nagyrobi** [reply]
I upvoted!

**Reply by fvanroie** [reply]
It is possible to move to the new core with some tweaks in the ini file:
https://github.com/platformio/platform-espressif32/issues/619#issuecomment-944241822

Add this to your environment to get ESP-IDF 4.4-dev:
```
platform  = https://github.com/platformio/platform-espressif32.git#feature/arduino-upstream
platform_packages =
   framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#2.0.0
```

![image](https://user-images.githubusercontent.com/15969459/137570320-3e204122-4fe6-4455-94cd-c0a7cc83d8e3.png)

**Reply by nagyrobi** [reply]
Which files should I change?

**Reply by fvanroie** [reply]
You need to insert that piece into the [user_setups\esp32\lanbon_l8.ini](https://github.com/HASwitchPlate/openHASP/blob/cf53d657b2d37b0e7f9e247baed770914c38ef0c/user_setups/esp32/lanbon_l8.ini#L14)

I'm testing this override per 1 evironment, as long as it is not officially released I'll leave the default to the stable core.

**Reply by nagyrobi** [reply]
Testing this on Lanbon. Generated OTA binary is slightly bigger.

**Reply by fvanroie** [reply]
In 2.0.0 memory consumption is not optimized yet, but Arduino 2.0.1 promises memory improvements: 
- Smaller memory footprint (smaller than even v1.0.6)

I noticed about 10kB less free heap on 2.0.0 but this will hopefully soon be fixed.

**Reply by fvanroie** [reply]
Updated v0.7.0-dev to Arduino 2.0.2. FragAttack should be fixed using this current version of the framework.
